## Torch activation and wget files

In [ ]:
# активация cuda 
import torch
# cuda0 = torch.device('cuda:0')
# device = torch.device("cpu")
device = torch.device("cuda:0" if torch.cuda.is_available() else "cpu")
print('Torch', torch.__version__, 'CUDA', torch.version.cuda)
print('Device:', torch.device('cuda:0'))
torch.cuda.get_device_name(0)

Torch 1.8.1+cu101 CUDA 10.1
Device: cuda:0


'Tesla T4'

In [11]:
!wget --no-check-certificate 'https://docs.google.com/uc?export=download&id=1vtRjwycDt0M773YEcknrk4SPg2Rgv7g1' -O qdq_slice.сsv
!wget --no-check-certificate 'https://docs.google.com/uc?export=download&id=1RI846sdFyfUWR7983olonCnvneFWYt3h' -O qdq_emb
!wget --no-check-certificate 'https://docs.google.com/uc?export=download&id=1M5NANl00_KY7bwlUACprUEDi9uQNcBoo' -O qdq_role_emb
# !wget --no-check-certificate 'https://docs.google.com/uc?export=download&id=1TsMGnk1P1sXPs88eUWZKHKzzR7fNk8ol' -O paraphrases_framebank.tsv
# !wget --no-check-certificate 'https://docs.google.com/uc?export=download&id=1_QGFNWIOj8jbqOwAcXM4ejQ1vzENy3Sj' -O paraphrases_all_framebank.txt
# !wget --no-check-certificate 'https://docs.google.com/uc?export=download&id=1FhZfl-fNIthFn5YBQxLutMn2rWlv9OXH' -O quora_question_pairs_rus.csv
# !wget --no-check-certificate 'https://docs.google.com/uc?export=download&id=13fCtGg1eiHyoChJNVahCaU407QE88JW5' -O qdq_slice.tsv
# !wget --load-cookies /tmp/cookies.txt "https://docs.google.com/uc?export=download&confirm=$(wget --quiet --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=16l3cvhNDmrFs-7FCikIfGyPUgs_GgnMt' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')&id=16l3cvhNDmrFs-7FCikIfGyPUgs_GgnMt" -O quora_duplicate_questions.tsv && rm -rf /tmp/cookies.txt
# !wget --load-cookies /tmp/cookies.txt "https://docs.google.com/uc?export=download&confirm=$(wget --quiet --save-cookies /tmp/cookies.txt --keep-session-cookies --no-check-certificate 'https://docs.google.com/uc?export=download&id=1miYAsOQO0lmdxUtYXC84nqVXck7Xf-nk' -O- | sed -rn 's/.*confirm=([0-9A-Za-z_]+).*/\1\n/p')&id=1miYAsOQO0lmdxUtYXC84nqVXck7Xf-nk" -O glove.840B.300d.sst.txt && rm -rf /tmp/cookies.txt
# !wget --no-check-certificate 'https://docs.google.com/uc?export=download&id=1BHA6WLG7O42r8va2QLH2isEnyT3VWLaL' -O paraphrases_gold.tsv
# !wget --no-check-certificate 'https://docs.google.com/uc?export=download&id=1LygNEEIWFFAHrvfl1UrrJ-1uJxA5ewZJ' -O paraphrases.tsv
# !wget --no-check-certificate 'https://docs.google.com/uc?export=download&id=1ika4aMrTUZaPrDCwRou0IqVqOKJF7KMO' -O paraphrases_emb

--2021-06-03 10:36:14--  https://docs.google.com/uc?export=download&id=1vtRjwycDt0M773YEcknrk4SPg2Rgv7g1
Resolving docs.google.com (docs.google.com)... 142.251.33.206, 2607:f8b0:4004:837::200e
Connecting to docs.google.com (docs.google.com)|142.251.33.206|:443... connected.
HTTP request sent, awaiting response... 302 Moved Temporarily
Location: https://doc-10-bo-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/3jd4lqtnq3dugq7me3t3v627hp781u8a/1622716575000/01559997439925083741/*/1vtRjwycDt0M773YEcknrk4SPg2Rgv7g1?e=download [following]
--2021-06-03 10:36:16--  https://doc-10-bo-docs.googleusercontent.com/docs/securesc/ha0ro937gcuc7l7deffksulhg5h7mbp1/3jd4lqtnq3dugq7me3t3v627hp781u8a/1622716575000/01559997439925083741/*/1vtRjwycDt0M773YEcknrk4SPg2Rgv7g1?e=download
Resolving doc-10-bo-docs.googleusercontent.com (doc-10-bo-docs.googleusercontent.com)... 172.217.2.97, 2607:f8b0:4004:80a::2001
Connecting to doc-10-bo-docs.googleusercontent.com (doc-10-bo-docs.googleu

In [ ]:
# !wget http://files.deeppavlov.ai/embeddings/ft_native_300_ru_wiki_lenta_lemmatize/ft_native_300_ru_wiki_lenta_lemmatize.vec
# !wget http://wikipedia2vec.s3.amazonaws.com/models/ru/2018-04-20/ruwiki_20180420_300d.txt.bz2
# !bzip2 -d ruwiki_20180420_300d.txt.bz2
!wget --output-document=- \
   http://wikipedia2vec.s3.amazonaws.com/models/ru/2018-04-20/ruwiki_20180420_300d.txt.bz2 \
  | bunzip2 > ruwiki_20180420_300d.txt

--2021-06-03 10:10:48--  http://wikipedia2vec.s3.amazonaws.com/models/ru/2018-04-20/ruwiki_20180420_300d.txt.bz2
Resolving wikipedia2vec.s3.amazonaws.com (wikipedia2vec.s3.amazonaws.com)... 52.219.68.37
Connecting to wikipedia2vec.s3.amazonaws.com (wikipedia2vec.s3.amazonaws.com)|52.219.68.37|:80... connected.
HTTP request sent, awaiting response... 200 OK
Length: 1457741710 (1.4G) [application/x-bzip2]
Saving to: ‘STDOUT’

-                   100%[===================>]   1.36G  4.46MB/s    in 5m 1s   

2021-06-03 10:15:50 (4.61 MB/s) - written to stdout [1457741710/1457741710]



In [ ]:
import random
import timeit
import numpy
import csv

filename = 'quora_question_pairs_rus.csv'
output_file = 'qdq_slice.tsv'

def talktime(filename, funcname, func):
    print(f"# {funcname}")
    t = timeit.timeit(f'{funcname}("{filename}")', setup=f'from __main__ import {funcname}', number = 100) / 100
    print('Elapsed time : ', t)
    print('n = ', func(filename))
    print('\n')

def sum1forline(filename):
    with open(filename) as f:
        return sum(1 for line in f)

talktime(filename, 'sum1forline', sum1forline)
file_size = sum1forline(filename)

with open(filename, encoding='utf-8') as f:   
  fieldnames = ['id','qid1','qid2','question1','question2','is_duplicate']
  delimiter = ','
  quotechar = ''        
  quoting = quoting=csv.QUOTE_NONE
  # fieldnames=fieldnames,        
  reader = csv.DictReader(f,fieldnames=fieldnames, delimiter=delimiter, quoting=quoting, quotechar=quotechar, escapechar='')

  data = list(reader)
  data = numpy.asarray(data)
  numpy.random.seed(123)
  numpy.random.shuffle(data)
  length = data.shape[0]
        
  sliced_data = data[:int(0.04 * length)]
  for line in sliced_data:
    print(line)

  print(len(sliced_data))

with open(output_file, "w", encoding='utf-8') as f:  
  fieldnames = ['id','qid1','qid2','question1','question2','is_duplicate']
  delimiter = '\t'
  quotechar = ''        
  quoting = quoting=csv.QUOTE_NONE
  output_writer = csv.DictWriter(f, fieldnames=fieldnames, delimiter=delimiter, quoting=quoting, quotechar=quotechar, escapechar='')
  output_writer.writeheader()
  
  for row in sliced_data:
    output_writer.writerow({
      'id': row['id'],
      'qid1': row['qid1'],
      'qid2': row['qid2'],
      'question1': row['question1'],
      'question2': row['question2'],
      'is_duplicate': row['is_duplicate']
    })


Выходные данные были обрезаны до нескольких последних строк (5000).
OrderedDict([('id', '327434'), ('qid1', 'как демонизация из 500 и 1000 рупийских нот повлияет на значение inr против usd'), ('qid2', 'будет увеличиваться стоимость рупии после запрета на замещение 500 и 1000 рупий'), ('question1', '1'), ('question2', None), ('is_duplicate', None)])
OrderedDict([('id', '403253'), ('qid1', 'Каков ваш обзор программы macbook за 2016 год?'), ('qid2', 'что разработчики думают о новом macbook pro'), ('question1', '1'), ('question2', None), ('is_duplicate', None)])
OrderedDict([('id', '339359'), ('qid1', 'как я могу приготовить желе со свежим киви'), ('qid2', '"сколько времени вам нужно'), ('question1', ' чтобы приготовить 1000 г риса"'), ('question2', '0'), ('is_duplicate', None)])
OrderedDict([('id', '211698'), ('qid1', 'с чего сейчас лучше всего работает фондовый портфель'), ('qid2', '"что такое хорошее программное обеспечение для портфеля ценных бумаг'), ('question1', ' которое я могу исп

IOPub data rate exceeded.
The notebook server will temporarily stop sending output
to the client in order to avoid crashing it.
To change this limit, set the config variable
`--NotebookApp.iopub_data_rate_limit`.

Current values:
NotebookApp.iopub_data_rate_limit=1000000.0 (bytes/sec)
NotebookApp.rate_limit_window=3.0 (secs)



OrderedDict([('id', '145471'), ('qid1', '"что лучший способ узнать Matlab'), ('qid2', ' что является лучшей книгой для изучения Matlab"'), ('question1', '"как узнать'), ('question2', ' что такое Matlab"'), ('is_duplicate', '0')])
OrderedDict([('id', '296038'), ('qid1', 'что такое английское слово для слова tagalog kinikilig'), ('qid2', 'что такое выражение tagalog для меня извините'), ('question1', '0'), ('question2', None), ('is_duplicate', None)])
OrderedDict([('id', '256713'), ('qid1', 'какова разница между аутизмом и умственно отсталыми пациентами'), ('qid2', 'каковы различия между аутизмом и adhd'), ('question1', '0'), ('question2', None), ('is_duplicate', None)])
OrderedDict([('id', '364434'), ('qid1', 'какой самый успокаивающий алкоголь'), ('qid2', '"это бар для найма кандидатов в промышленность'), ('question1', ' выше в техногенных компаниях"'), ('question2', '0'), ('is_duplicate', None)])
OrderedDict([('id', '357755'), ('qid1', 'почему люди не могут понять чувства'), ('qid2', 

## Preprocess rus data (create data_emb from rus. dataset)

In [10]:
!pip install razdel
!pip install natasha

import pickle
import os
from razdel import tokenize, sentenize
import numpy
import csv
from natasha import (
    Segmenter,
    MorphVocab,
    NewsEmbedding,
    NewsMorphTagger,
    NewsSyntaxParser,
    NewsNERTagger,
    PER,
    NamesExtractor,
    DatesExtractor,
    MoneyExtractor,
    AddrExtractor,
    Doc
)
import gc
gc.enable()

segmenter = Segmenter()
morph_vocab = MorphVocab()

emb = NewsEmbedding()
morph_tagger = NewsMorphTagger(emb)
syntax_parser = NewsSyntaxParser(emb)
ner_tagger = NewsNERTagger(emb)

names_extractor = NamesExtractor(morph_vocab)
dates_extractor = DatesExtractor(morph_vocab)
money_extractor = MoneyExtractor(morph_vocab)
addr_extractor = AddrExtractor(morph_vocab)

# paraphrases_framebank.tsv
def read_all_data(train_file='qdq_slice.сsv'):
    with open(train_file, encoding='utf-8') as f:    
        # fieldnames = ['phrase1','roles1','phrase2','roles2','is_duplicate']
        delimiter = ','
        quotechar = ''        
        quoting = quoting=csv.QUOTE_NONE
        # fieldnames=fieldnames,        
        reader = csv.DictReader(f, delimiter=delimiter, quoting=quoting, quotechar=quotechar, escapechar='')

        data = list(reader)
        data = numpy.asarray(data)
        numpy.random.seed(123)
        numpy.random.shuffle(data)
        length = data.shape[0]
        
        train = data[:int(0.85 * length)]
        valid = data[int(0.85 * length):int(0.9 * length)]
        test = data[int(0.9 * length):]

        return train, valid, test

def _get_tokens_from_text(text,is_roles):
    if text is None:
        return []
    if is_roles:
        text = text.strip().replace('`', "'")
        tokens = text.split(',') 
    else:
        text = text.strip().replace('`', "'")
        tokens = text.split(' ') 
    
    return tokens


def split_data(all_data):
    tokenized = []
    uniq_words = set()

    for set_ in all_data:
        data = {'q1': [], 'rl1': [], 'q2': [], 'rl2': [], 'y': []}

        for datum in set_:
            try:
              data['y'].append(int(datum['is_duplicate']))
            except ValueError as e:
              print(datum)
              print(datum['is_duplicate'])
              print('error type: ', type (e))
              continue

            q1 = _get_tokens_from_text(datum['question1'],False)
            q2 = _get_tokens_from_text(datum['question2'],False)
            # rl1 = _get_tokens_from_text(datum['roles1'],True)
            # rl2 = _get_tokens_from_text(datum['roles2'],True)
            data['q1'].append(q1)
            # data['rl1'].append(rl1)
            data['q2'].append(q2)
            # data['rl2'].append(rl2)
            # -1 -> 0  0 -> 1  1 -> 2            
            # if len(q1) != len(rl1):
            #   print("ROW1:")
            #   print(len(q1))
            #   print(q1)
            #   print(datum['phrase1'])
            #   print(len(rl1))
            #   print(rl1)
            #   print(datum['roles1'])
            # if len(q2) != len(rl2):
            #   print("ROW2:")
            #   print(len(q2))
            #   print(q2)
            #   print(datum['phrase2'])
            #   print(len(rl2))
            #   print(rl2)
            #   print(datum['roles2'])

            for token in q1:
              if token not in uniq_words:
                uniq_words.add(token)

            for token in q2:
              if token not in uniq_words:
                uniq_words.add(token)
                            
        tokenized.append(data)

    return tokenized, uniq_words

# def _find_emb(file,dim,token):
#     dim += 1
#     with open(file, encoding='utf-8') as f:
#         for line in f:
#             tokens = line.strip().split(' ')
#             if len(tokens) == dim and tokens[0] == token:
#                 return list(map(lambda x: float(x), tokens[1:]))
#                 # vect = 
#                 # emb[tokens[0]] = list(map(lambda x: float(x), tokens[1:]))
#     return False

def _read_emb(file, dim, uniq_words):
    emb = {}
    dim += 1
    print(len(uniq_words))
    with open(file, encoding='utf-8') as f:
        for line in f:
            tokens = line.strip().split(' ')
            if len(tokens) == dim and tokens[0] in uniq_words:
                uniq_words.discard(tokens[0])
                emb[tokens[0]] = list(map(lambda x: float(x), tokens[1:]))
                # print(emb[tokens[0]])
                
    print(len(uniq_words))
    return emb

def _token2idx(tokens, token_map, embs, filtered_emb):
    for i in range(len(tokens)):
        if tokens[i] not in token_map:
            if tokens[i] in embs:
                token_map[tokens[i]] = len(token_map)
                filtered_emb.append(embs[tokens[i]])
            else:
                tokens[i] = '<unk>'
        tokens[i] = token_map[tokens[i]]

def idx_and_emb(all_data, emb_file, dim, uniq_words):
    embs = _read_emb(emb_file, dim, uniq_words)

    word2idx = {'<pad>': 0, '<unk>': 1}
    filtered_emb = [numpy.random.uniform(-0.1, 0.1, dim) for _ in range(2)]

    for set_ in all_data:
        for datum in set_['q1']:
            _token2idx(datum, word2idx, embs, filtered_emb)
        for datum in set_['q2']:
            _token2idx(datum, word2idx, embs, filtered_emb)
    print('{} word types'.format(len(word2idx)))
    filtered_emb = numpy.asarray(filtered_emb, dtype='float32')
    return filtered_emb, word2idx

def _role_token2idx(tokens, token_map, role_filtered_emb, dim):
    for i in range(len(tokens)):
        if tokens[i] not in token_map:
            token_map[tokens[i]] = len(token_map)
            role_filtered_emb.append(numpy.random.uniform(-0.1, 0.1, dim))
        tokens[i] = token_map[tokens[i]]

def idx_and_emb_with_role(all_data, dim):
    role_word2idx = {'<pad>': 0, '<unk>': 1, '<pred>': 2}
    role_filtered_emb = [numpy.random.uniform(-0.1, 0.1, dim) for _ in range(3)]    

    for set_ in all_data:
        for datum in set_['rl1']:
            _role_token2idx(datum, role_word2idx, role_filtered_emb, dim)
        for datum in set_['rl2']:
            _role_token2idx(datum, role_word2idx, role_filtered_emb, dim)

    print('{} roles types'.format(len(role_word2idx)))
    role_filtered_emb = numpy.asarray(role_filtered_emb, dtype='float32')
    return role_filtered_emb, role_word2idx

def main():
    all_data, uniq_words = split_data(read_all_data())
    print('Reading data done.')
    # emb_file = 'glove.840B.300d.sst.txt'
    emb_file = 'ruwiki_20180420_300d.txt'

    filtered_emb, word2idx = idx_and_emb(all_data, emb_file, 300, uniq_words)
    print('Embedding done.')

    # role_filtered_emb, role_word2idx = idx_and_emb_with_role(all_data, 50)
    # print('Role embedding done.')    

    with open('paraphrases_emb', 'wb') as f:
        pickle.dump((all_data, filtered_emb, word2idx), f)
    print('Saved.')

main()

# from google.colab import files
# files.download('paraphrases_emb') 

Выходные данные были обрезаны до нескольких последних строк (5000).
как вы сделали свой первый миллион долларов
error type:  <class 'ValueError'>
OrderedDict([('id', '376779'), ('question1', 'почему холмик продолжает пытаться убить Assange'), ('question2', '"Хиллари Клинтон предложила ударить беспилотник'), ('is_duplicate', ' чтобы убить юлианский союз"'), (None, ['1'])])
 чтобы убить юлианский союз"
error type:  <class 'ValueError'>
OrderedDict([('id', '321861'), ('question1', '"есть ли шанс'), ('question2', ' что я смогу найти подругу'), ('is_duplicate', ' когда у меня возникнет синдром аспергера'), (None, [' самое худшее расстройство человечества"', '"есть ли какая-нибудь надежда найти подругу', ' если я стесняюсь синдрома аспергера', ' и я серьезно"', '1'])])
 когда у меня возникнет синдром аспергера
error type:  <class 'ValueError'>
OrderedDict([('id', '320416'), ('question1', 'как часто человеку приходится простужаться через 1 год'), ('question2', '"вы можете болеть только от хол

## Utils for model work

In [ ]:
import numpy
import os

def pad(var_arr, max_len=None, fix_len=False):
    if max_len is None:
        seq_len = [len(x) for x in var_arr]
        max_len = max(seq_len)
    else:
        assert max_len > 0
        seq_len = [min(len(x), max_len) for x in var_arr]
        if not fix_len:
            max_len = max(seq_len)
    fixed_var = numpy.zeros((len(var_arr), max_len), dtype=numpy.int64)
    for idx, x in enumerate(var_arr):
        fixed_var[idx][0:seq_len[idx]] = x[0:seq_len[idx]]
    for i in range(len(seq_len)):
        if seq_len[i] < 1:
            seq_len[i] = 1
    return fixed_var, seq_len

# def to_categorical(y, num_classes):
#     """ 1-hot encodes a tensor """
#     return numpy.eye(num_classes, dtype='uint8')[y]

def pad2d(var_arr):
    max_len1 = max_len2 = 0
    for arrs in var_arr:
        max_len1 = max(len(arrs), max_len1)
        for arr in arrs:
            max_len2 = max(len(arr), max_len2)
    fixed_arr = numpy.zeros((len(var_arr), max_len1, max_len2), dtype=numpy.int64)
    for i, arrs in enumerate(var_arr):
        for j, arr in enumerate(arrs):
            for k, val in enumerate(arr):
                fixed_arr[i, j, k] = val
    return fixed_arr


def pad_vec(var_arr):  # Sparse matrix
    vec_len = var_arr[0].shape[-1]
    assert vec_len != 0
    seq_len = [x.shape[0] for x in var_arr]
    max_len = max(seq_len)
    fixed_var = numpy.zeros((len(var_arr), max_len, vec_len), dtype=numpy.float32)
    for i, x in enumerate(var_arr):
        for j, k in zip(*x.nonzero()):
            fixed_var[i, j, k] = x[j, k]
    for i in range(len(seq_len)):
        if seq_len[i] < 1:
            seq_len[i] = 1
    return fixed_var, seq_len


class Dataset:
    def __init__(self, data, shuffle=False, pad_keys=(), pad2d_keys=()):
        self._data = {key: numpy.array(val) for key, val in data.items()}
        self._epochs_completed = 0
        self._num_examples = next(iter(self._data.values())).shape[0]
        self._index = numpy.arange(self._num_examples)
        self._shuffle = shuffle
        self._batches = []
        self._pad_keys = pad_keys
        self._seq_lens = []
        self._pad2d_keys = pad2d_keys

    def epochs_completed(self):
        return self._epochs_completed

    def get_batches(self, batch_size, keys):
        if self._shuffle or not self._batches:
            if self._shuffle:
                numpy.random.shuffle(self._index)
            self._build_batches(batch_size, keys)
        self._epochs_completed += 1
        return self._batches, self._seq_lens

    def _build_batches(self, batch_size, keys):
        self._batches = []
        self._seq_lens = []
        for start in range(0, self._num_examples, batch_size):
            batch = {}
            batch_seq_lens = {}
            for key in keys:
                batch_val = self._data[key][self._index[start:start + batch_size]]
                if key in self._pad_keys:
                    batch_val, seq_lens_val = pad(batch_val)
                    batch_seq_lens[key] = seq_lens_val
                elif key in self._pad2d_keys:
                    batch_val = pad2d(batch_val)
                batch[key] = batch_val
            self._batches.append(batch)
            self._seq_lens.append(batch_seq_lens)

    def get_singles(self, keys):
        if self._shuffle or not self._batches:
            if self._shuffle:
                numpy.random.shuffle(self._index)
            self._build_batches(1, keys)
        for single in self._batches:
            for key, val in single.items():
                single[key] = val[0]
        self._epochs_completed += 1
        return self._batches

    def get_data_item(self, key):
        return self._data[key]


def mask(lengths):
    lengths = numpy.array(lengths)
    lengths = lengths.reshape(lengths.shape[0], -1)
    max_lens = lengths.max(axis=0)
    batch_size = lengths.shape[0]
    mask_size = [batch_size] + max_lens.tolist()
    masks = numpy.zeros(mask_size, dtype=numpy.float32)
    num_dim = len(max_lens)
    assert num_dim < 3
    for sample, length in zip(masks, lengths):
        if num_dim == 1:
            sample[:length[0]] = 1
        elif num_dim == 2:
            sample[:length[0], :length[1]] = 1
    return masks


def copy_file(ori_file, dest_file):
    with open(ori_file) as fin, open(dest_file, 'w') as fout:
        for line in fin:
            fout.write(line)


def get_script_short_name(name):
    name = os.path.basename(name)
    name = name[:name.find('.')]
    return name


class Accuracy:
    def __init__(self):
        self.total = 0
        self.count = 0

    def update_batch(self, truth, pred):
        assert len(truth) == len(pred)
        for t, p in zip(truth, pred):
            if t == p:
                self.count += 1
        self.total += len(truth)

    def get_score(self):
        return self.count / self.total

# Model's

##Model's on eng. data

Если запустить preprocess rus data, то на русских данных

##AI-BLSTM NN (87.43 baseline accuracy) - ...

1.   У меня вышло 84.30 ... Best model on dev: dev:0.8458	test:0.8430


In [ ]:
!touch AI-BLSTM.txt
!touch AI-BLSTM.pred
!ls .

AI-BLSTM.pred  glove.840B.300d.sst.txt	paraphrases.tsv
AI-BLSTM.txt   paraphrases_emb		quora_duplicate_questions.tsv
drive	       paraphrases_gold.tsv	sample_data


In [ ]:
import time
import torch
from torch import nn
from torch.nn import functional as func
from torch.autograd import Variable
from torch.nn.utils import rnn
import pickle
import numpy
from sklearn.metrics import accuracy_score
# from utils import Dataset, get_script_short_name

hparams = {
    'learning_rate': 0.001,
    'max_epoch': 10,
    # 1000 for eng,
    'display_step': 1,
    'emb_dim': 300,
    'conv_win': 3,
    'repr_dim': 300,
    'fc0_dim': 300,
    'fc1_dim': 300,
    # для русского 3, для eng 2
    'n_classes': 2,
    'batch_size': 100
}


class LSTM(nn.Module):
    def __init__(self, emb_layer):
        super(LSTM, self).__init__()
        self.emb_layer = emb_layer
        self.lstm_layer = nn.LSTM(input_size=hparams['emb_dim'], hidden_size=int(hparams['repr_dim'] / 2),
                                  batch_first=True, bidirectional=True)

    def forward(self, x, lengths):
        embs = self.emb_layer(x)
        embs_sort, lengths_sort, unsort_idx = self.sort_batch(embs, lengths)
        seq = rnn.pack_padded_sequence(embs_sort, lengths_sort.cpu().numpy(), batch_first=True)
        hs, _ = self.lstm_layer(seq)
        hs, _ = rnn.pad_packed_sequence(hs, batch_first=True)
        return hs[unsort_idx]

    @staticmethod
    def sort_batch(x, l):
      # .cuda()
        l = torch.from_numpy(numpy.asarray(l)).cuda()
        l_sorted, sidx = l.sort(0, descending=True)
        x_sorted = x[sidx]
        _, unsort_idx = sidx.sort()
        return x_sorted, l_sorted, unsort_idx


class Model(nn.Module):
    def __init__(self, emb_layer):
        super(Model, self).__init__()
        self.lstm = LSTM(emb_layer)
        self.fc0 = nn.Linear(hparams['repr_dim'] * 2, hparams['fc0_dim'])
        self.fc1 = nn.Linear(hparams['fc0_dim'] * 2, hparams['fc1_dim'])
        self.fc2 = nn.Linear(hparams['fc1_dim'], hparams['n_classes'])
        self.att = nn.Linear(hparams['repr_dim'] + hparams['fc0_dim'], 1)

    def forward(self, q1, q2, q1_len, q2_len):
        r1 = self.lstm(q1, q1_len)
        r2 = self.lstm(q2, q2_len)
        
        h1s, h2s = [], []
        
        for i in range(len(q1_len)):
            
            r1i = r1[i, :q1_len[i]]
            r2i = r2[i, :q2_len[i]]
            
            r1i_size = r1i.size()
            r2i_size = r2i.size()
            
            joint_size = torch.Size((r1i_size[0], r2i_size[0], r2i_size[1]))
            
            r1ie = r1i.unsqueeze(1).expand(joint_size)
            r2ie = r2i.unsqueeze(0).expand(joint_size)
            
            joint = torch.cat([(r1ie - r2ie).abs(), r2ie * r1ie], 2)
            joint = torch.tanh(self.fc0(joint.view(-1, joint.size()[-1])))
            #  func.tanh()
            joint = joint.view(joint_size[0], joint_size[1], hparams['fc0_dim'])

            s1, _ = joint.max(1)
            s2, _ = joint.max(0)
        
            s1 = s1.squeeze(1)
            s2 = s2.squeeze(0)

            if (len(s2.shape) == 1):
                s2 = torch.reshape(s2,(1,300))

            a1 = self.att(torch.cat([r1i, s1], 1))
            a2 = self.att(torch.cat([r2i, s2], 1))
        
            a1 = func.softmax(a1.t(),dim=1).t().expand_as(s1)
            a2 = func.softmax(a2.t(),dim=1).t().expand_as(s2)
        
            h1 = (s1 * a1).sum(0).squeeze()
            h2 = (s2 * a2).sum(0).squeeze()
        
            h1s.append(h1)
            h2s.append(h2)
        
        r1 = torch.stack(h1s)
        r2 = torch.stack(h2s)
        
        joint = torch.cat((r1, r2), dim=1)
        joint = torch.tanh(self.fc1(joint))
        
        out = self.fc2(joint)
        return out


def run_batch(b_data, b_lengths, model, optimizer=None):
    # .cuda()
    q1 = Variable(torch.from_numpy(b_data['q1'])).cuda()
    q2 = Variable(torch.from_numpy(b_data['q2'])).cuda()
    
    outputs = model(q1, q2, b_lengths['q1'], b_lengths['q2'])
    
    if optimizer:
        entropy_loss = nn.CrossEntropyLoss()
        target = Variable(torch.from_numpy(b_data['y'])).cuda()
        optimizer.zero_grad()
        
        # print(target)    
        # print(outputs)
        # print(outputs.data)
        # print(func.softmax(outputs).data)
        
        # loss = func.cross_entropy(outputs, y)
        loss = entropy_loss(outputs, target)
        loss.backward()
        
        optimizer.step()
    
        return loss.item()
        # return loss.data[0]
        # return loss.data
    else:
        _, predicted = outputs.data.max(1)
        prob = func.softmax(outputs).data
        # print(prob)
        # print(predicted)
        return predicted, prob[:, 1]

# 
def run_epoch_eval(dataset, model, output_file=''):
    all_plabels, all_pscores = [], []
    batches, batch_lengths = dataset.get_batches(hparams['batch_size'], ('q1', 'q2', 'y'))
    for b_data, b_lengths in zip(batches, batch_lengths):
        plabels, pscores = run_batch(b_data, b_lengths, model)
        all_plabels.extend(plabels.cpu().numpy().tolist())
        all_pscores.extend(pscores.cpu().numpy().tolist())
    if output_file:
        with open(output_file, 'w') as f:
            for s in all_pscores:
                f.write(f'{s:.6f}\n')
    return accuracy_score(dataset.get_data_item('y'), all_plabels)


def run():
    print('Loading data...')
    with open('paraphrases_emb', 'rb') as f:
      all_sets, embs, word2idx = pickle.load(f)
    
    emb_layer = nn.Embedding(embs.shape[0], embs.shape[1])
    emb_layer.weight = nn.Parameter(torch.from_numpy(embs))
    
    # role_emb_layer = 
    # role_emb_layer.weight =
    # print(emb_layer)
    # print(embs.shape[0])
    # print(embs.shape[1])
    # print(embs)

    model = Model(emb_layer).cuda()
    # .cuda()
    model = model.to(device)

    optimizer = torch.optim.Adam(model.parameters(), lr=hparams['learning_rate'])
    
    train_set = Dataset(all_sets[0], shuffle=True, pad_keys=('q1', 'q2'))
    dev_set = Dataset(all_sets[1], shuffle=False, pad_keys=('q1', 'q2'))
    test_set = Dataset(all_sets[2], shuffle=False, pad_keys=('q1', 'q2'))

    step = 0
    sum_loss = 0
    dev_best = 0
    test_score = 0
    
    print("Starting training...")
    print(hparams)
    
    start_time = time.time()
    
    for epoch in range(hparams['max_epoch']):
    
        batches, batch_lengths = train_set.get_batches(hparams['batch_size'], ('q1', 'q2', 'y'))
        
        for b_data, b_lengths in zip(batches, batch_lengths):
    
            sum_loss += run_batch(b_data, b_lengths, model, optimizer)
            step += 1
          
        avg_loss = sum_loss / len(batches)
        sum_loss = 0
        dev_score = run_epoch_eval(dev_set, model)
        out_str = f'Epoch {epoch} iter {step} took {time.time() - start_time:.1f}s\n' \
                  f'loss:\t{avg_loss:.5f}\tdev score:\t{dev_score:.4f}'

        if dev_score > dev_best:
          dev_best = dev_score
          dev_best_epoch = epoch
          # output_file = f'pred/{get_script_short_name(__file__)}.pred'
          test_score = run_epoch_eval(test_set, model, 'AI-BLSTM')
          out_str += f'\t*** New best dev ***\ttest score:\t{test_score:.4f}'
    
        print(out_str)
        start_time = time.time()
        
    print('Best model on dev: dev:{:.4f}\ttest:{:.4f}'.format(dev_best, test_score))
    print('On {:.4f} epoch'.format(dev_best_epoch))

if __name__ == '__main__':
    run()

Loading data...
Starting training...
{'learning_rate': 0.001, 'max_epoch': 10, 'display_step': 1, 'emb_dim': 300, 'conv_win': 3, 'repr_dim': 300, 'fc0_dim': 300, 'fc1_dim': 300, 'n_classes': 2, 'batch_size': 100}


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:56: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:145: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.


Epoch 0 iter 213 took 375.4s
loss:	0.56491	dev score:	0.7454	*** New best dev ***	test score:	0.7287
Epoch 1 iter 426 took 372.1s
loss:	0.39698	dev score:	0.8183	*** New best dev ***	test score:	0.7931
Epoch 2 iter 639 took 364.2s
loss:	0.23923	dev score:	0.8679	*** New best dev ***	test score:	0.8427
Epoch 3 iter 852 took 367.5s
loss:	0.13006	dev score:	0.8903	*** New best dev ***	test score:	0.8523
Epoch 4 iter 1065 took 370.8s
loss:	0.07820	dev score:	0.8703
Epoch 5 iter 1278 took 370.4s
loss:	0.05978	dev score:	0.8719
Epoch 6 iter 1491 took 370.5s
loss:	0.04243	dev score:	0.8887
Epoch 7 iter 1704 took 363.6s
loss:	0.02925	dev score:	0.8863
Epoch 8 iter 1917 took 362.0s
loss:	0.03161	dev score:	0.8687
Epoch 9 iter 2130 took 363.7s
loss:	0.02881	dev score:	0.8895
Best model on dev: dev:0.8903	test:0.8523
On 3.0000 epoch


## BILSTM (86.17) 

- Best model on dev: dev:0.8408	test:0.8389

In [ ]:
!touch BLSTM.txt
!touch BLSTM.pred
!ls .

BLSTM.pred  CNN.txt		     quora_duplicate_questions.tsv
BLSTM.txt   data_emb		     quora_question_pairs_rus.csv
CNN	    data_emb.txt	     sample_data
CNN.pred    glove.840B.300d.sst.txt


In [ ]:
import time
import torch
from torch import nn
from torch.nn import functional as func
from torch.nn.utils import rnn
from torch.autograd import Variable
import numpy
import pickle
from sklearn.metrics import accuracy_score
# from utils import Dataset, get_script_short_name
 
hparams = {
    'learning_rate': 0.001,
    'max_epoch': 10,
    'display_step': 1000,
    'emb_dim': 300,
    'conv_win': 3,
    'repr_dim': 300,
    'fc1_dim': 300,
    'n_classes': 2,
    'batch_size': 100
}
 
 
class LSTM(nn.Module):
    def __init__(self, emb_layer):
        super(LSTM, self).__init__()
        self.emb_layer = emb_layer
        self.lstm_layer = nn.LSTM(input_size=hparams['emb_dim'], hidden_size=int(hparams['repr_dim'] / 2),
                                  batch_first=True, bidirectional=True)
 
    def forward(self, x, lengths):
        embs = self.emb_layer(x)
        embs_sort, lengths_sort, unsort_idx = self.sort_batch(embs, lengths)
        seq = rnn.pack_padded_sequence(embs_sort, lengths_sort.cpu().numpy(), batch_first=True)
        hs, (hn, cn) = self.lstm_layer(seq)
        out = cn.permute(1, 2, 0).contiguous().view(-1, hparams['repr_dim'])
        return out[unsort_idx]
 
    @staticmethod
    def sort_batch(x, l):
        # .cuda()
        l = torch.from_numpy(numpy.asarray(l)).cuda()
        l_sorted, sidx = l.sort(0, descending=True)
        x_sorted = x[sidx]
        _, unsort_idx = sidx.sort()
        return x_sorted, l_sorted, unsort_idx
 
 
class Model(nn.Module):
    def __init__(self, emb_layer):
        super(Model, self).__init__()
        self.lstm = LSTM(emb_layer)
        self.fc1 = nn.Linear(hparams['repr_dim'] * 2, hparams['fc1_dim'])
        self.fc2 = nn.Linear(hparams['fc1_dim'], hparams['n_classes'])
 
    def forward(self, q1, q2, q1_len, q2_len):
        r1 = self.lstm(q1, q1_len)
        r2 = self.lstm(q2, q2_len)
        joint = torch.cat(((r1 - r2).abs(), r1 * r2), dim=1)
        joint = func.tanh(self.fc1(joint))
        out = self.fc2(joint)
        return out
 
 
def run_batch(b_data, b_lengths, model, optimizer=None):
    # .cuda()
    q1 = Variable(torch.from_numpy(b_data['q1'])).cuda()
    q2 = Variable(torch.from_numpy(b_data['q2'])).cuda()
    
    outputs = model(q1, q2, b_lengths['q1'], b_lengths['q2'])
    
    if optimizer:
        entropy_loss = nn.CrossEntropyLoss()
        target = Variable(torch.from_numpy(b_data['y'])).cuda()
        optimizer.zero_grad()
        
        # print(target)    
        # print(outputs)
        # print(outputs.data)
        # print(func.softmax(outputs).data)
        
        # loss = func.cross_entropy(outputs, y)
        loss = entropy_loss(outputs, target)
        loss.backward()
        
        optimizer.step()
    
        return loss.item()
        # return loss.data[0]
        # return loss.data
    else:
        _, predicted = outputs.data.max(1)
        prob = func.softmax(outputs).data
        # print(prob)
        # print(predicted)
        return predicted, prob[:, 1]


def run_epoch_eval(dataset, model, output_file=''):
    all_plabels, all_pscores = [], []
    batches, batch_lengths = dataset.get_batches(hparams['batch_size'], ('q1', 'q2', 'y'))
    for b_data, b_lengths in zip(batches, batch_lengths):
        plabels, pscores = run_batch(b_data, b_lengths, model)
        all_plabels.extend(plabels.cpu().numpy().tolist())
        all_pscores.extend(pscores.cpu().numpy().tolist())
    if output_file:
        with open(output_file, 'w') as f:
            for s in all_pscores:
                f.write(f'{s:.6f}\n')
    return accuracy_score(dataset.get_data_item('y'), all_plabels)


def run():
    print('Loading data...')
    with open('paraphrases_emb', 'rb') as f:
      all_sets, embs, word2idx = pickle.load(f)
    
    emb_layer = nn.Embedding(embs.shape[0], embs.shape[1])
    emb_layer.weight = nn.Parameter(torch.from_numpy(embs))
    
    # role_emb_layer = 
    # role_emb_layer.weight =
    # print(emb_layer)
    # print(embs.shape[0])
    # print(embs.shape[1])
    # print(embs)

    model = Model(emb_layer).cuda()
    # .cuda()
    model = model.to(device)

    optimizer = torch.optim.Adam(model.parameters(), lr=hparams['learning_rate'])
    
    train_set = Dataset(all_sets[0], shuffle=True, pad_keys=('q1', 'q2'))
    dev_set = Dataset(all_sets[1], shuffle=False, pad_keys=('q1', 'q2'))
    test_set = Dataset(all_sets[2], shuffle=False, pad_keys=('q1', 'q2'))

    step = 0
    sum_loss = 0
    dev_best = 0
    test_score = 0
    
    print("Starting training...")
    print(hparams)
    
    start_time = time.time()
    
    for epoch in range(hparams['max_epoch']):
    
        batches, batch_lengths = train_set.get_batches(hparams['batch_size'], ('q1', 'q2', 'y'))
        
        for b_data, b_lengths in zip(batches, batch_lengths):
    
            sum_loss += run_batch(b_data, b_lengths, model, optimizer)
            step += 1
          
        avg_loss = sum_loss / len(batches)
        sum_loss = 0
        dev_score = run_epoch_eval(dev_set, model)
        out_str = f'Epoch {epoch} iter {step} took {time.time() - start_time:.1f}s\n' \
                  f'loss:\t{avg_loss:.5f}\tdev score:\t{dev_score:.4f}'

        if dev_score > dev_best:
          dev_best = dev_score
          dev_best_epoch = epoch
          # output_file = f'pred/{get_script_short_name(__file__)}.pred'
          test_score = run_epoch_eval(test_set, model, 'AI-BLSTM')
          out_str += f'\t*** New best dev ***\ttest score:\t{test_score:.4f}'
    
        print(out_str)
        start_time = time.time()
        
    print('Best model on dev: dev:{:.4f}\ttest:{:.4f}'.format(dev_best, test_score))
    print('On {:.4f} epoch'.format(dev_best_epoch))

if __name__ == '__main__':
    run()

Loading data...
Starting training...
{'learning_rate': 0.001, 'max_epoch': 10, 'display_step': 1000, 'emb_dim': 300, 'conv_win': 3, 'repr_dim': 300, 'fc1_dim': 300, 'n_classes': 2, 'batch_size': 100}


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:56: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
/usr/local/lib/python3.7/dist-packages/torch/nn/functional.py:1698: UserWarning: nn.functional.tanh is deprecated. Use torch.tanh instead.
  warnings.warn("nn.functional.tanh is deprecated. Use torch.tanh instead.")
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:94: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.


Epoch 0 iter 213 took 3.4s
loss:	0.58379	dev score:	0.7102	*** New best dev ***	test score:	0.6983
Epoch 1 iter 426 took 3.2s
loss:	0.40690	dev score:	0.7982	*** New best dev ***	test score:	0.8039
Epoch 2 iter 639 took 3.3s
loss:	0.21850	dev score:	0.8431	*** New best dev ***	test score:	0.8323
Epoch 3 iter 852 took 3.3s
loss:	0.11394	dev score:	0.8599	*** New best dev ***	test score:	0.8331
Epoch 4 iter 1065 took 3.2s
loss:	0.06697	dev score:	0.8663	*** New best dev ***	test score:	0.8551
Epoch 5 iter 1278 took 3.2s
loss:	0.04605	dev score:	0.8751	*** New best dev ***	test score:	0.8535
Epoch 6 iter 1491 took 3.3s
loss:	0.03775	dev score:	0.8775	*** New best dev ***	test score:	0.8599
Epoch 7 iter 1704 took 3.3s
loss:	0.02786	dev score:	0.8783	*** New best dev ***	test score:	0.8591
Epoch 8 iter 1917 took 3.3s
loss:	0.02094	dev score:	0.8767
Epoch 9 iter 2130 took 3.3s
loss:	0.02473	dev score:	0.8695
Best model on dev: dev:0.8783	test:0.8591
On 7.0000 epoch


## DA-BLSTM - (85.72 Baseline accuracy) 

Best model on dev: dev:0.8389	test:0.8370

In [12]:
# !touch DA-BLSTM.txt
# !touch DA-BLSTM.pred
# !ls .

import time
import torch
import numpy
from torch import nn
from torch.nn import functional as func
from torch.autograd import Variable
from torch.nn.utils import rnn
import pickle
from sklearn.metrics import accuracy_score
# from utils import Dataset, get_script_short_name

hparams = {
    'learning_rate': 0.001,
    'max_epoch': 10,
    'display_step': 1000,
    'emb_dim': 300,
    'conv_win': 3,
    'repr_dim': 300,
    'fc0_dim': 300,
    'fc1_dim': 300,
    'n_classes': 2,
    'batch_size': 100
}


class LSTM(nn.Module):
    def __init__(self, emb_layer):
        super(LSTM, self).__init__()
        self.emb_layer = emb_layer
        self.lstm_layer = nn.LSTM(input_size=hparams['emb_dim'], hidden_size=int(hparams['repr_dim'] / 2),
                                  batch_first=True, bidirectional=True)

    def forward(self, x, lengths):
        embs = self.emb_layer(x)
        embs_sort, lengths_sort, unsort_idx = self.sort_batch(embs, lengths)
        seq = rnn.pack_padded_sequence(embs_sort, lengths_sort.cpu().numpy(), batch_first=True)
        hs, _ = self.lstm_layer(seq)
        hs, _ = rnn.pad_packed_sequence(hs, batch_first=True)
        return hs[unsort_idx]

    @staticmethod
    def sort_batch(x, l):
        # .cuda()
        l = torch.from_numpy(numpy.asarray(l)).cuda()
        l_sorted, sidx = l.sort(0, descending=True)
        x_sorted = x[sidx]
        _, unsort_idx = sidx.sort()
        return x_sorted, l_sorted, unsort_idx


class Model(nn.Module):
    def __init__(self, emb_layer):
        super(Model, self).__init__()
        self.lstm = LSTM(emb_layer)
        self.fc0 = nn.Linear(hparams['repr_dim'] * 2, hparams['fc0_dim'])
        self.fc1 = nn.Linear(hparams['fc0_dim'] * 2, hparams['fc1_dim'])
        self.fc2 = nn.Linear(hparams['fc1_dim'], hparams['n_classes'])
        self.att = nn.Linear(hparams['repr_dim'] * 2, 1)

    def compare(self, r1, r2, g):
        s = func.softmax(g).unsqueeze(2)
        r2 = r2.unsqueeze(0)
        joint_size = s.size()[:-1] + r2.size()[-1:]
        s = s.expand(joint_size)
        r2 = r2.expand(joint_size)
        ar2 = (s * r2).sum(1).squeeze(1)
        comp = torch.cat([(r1 - ar2).abs(), ar2 * r1], 1)
        comp = func.tanh(self.fc0(comp))
        h = comp.mean(0).squeeze(0)
        return h

    def forward(self, q1, q2, q1_len, q2_len):
        r1 = self.lstm(q1, q1_len)
        r2 = self.lstm(q2, q2_len)
        h1s, h2s = [], []
        for i in range(len(q1_len)):
            r1i = r1[i, :q1_len[i]]
            r2i = r2[i, :q2_len[i]]
            r1i_size = r1i.size()
            r2i_size = r2i.size()
            joint_size = torch.Size((r1i_size[0], r2i_size[0], r2i_size[1]))
            r1ie = r1i.unsqueeze(1).expand(joint_size)
            r2ie = r2i.unsqueeze(0).expand(joint_size)
            joint = torch.cat([r1ie, r2ie], 2)
            g = func.tanh(self.att(joint.view(-1, joint.size()[-1])))
            g = g.view(joint_size[0], joint_size[1])
            h1 = self.compare(r1i, r2i, g)
            h2 = self.compare(r2i, r1i, g.t())
            h1s.append(h1)
            h2s.append(h2)
        r1 = torch.stack(h1s)
        r2 = torch.stack(h2s)
        joint = torch.cat((r1, r2), dim=1)
        joint = torch.tanh(self.fc1(joint))
        out = self.fc2(joint)
        return out


def run_batch(b_data, b_lengths, model, optimizer=None):
    # .cuda()
    q1 = Variable(torch.from_numpy(b_data['q1'])).cuda()
    q2 = Variable(torch.from_numpy(b_data['q2'])).cuda()
    
    outputs = model(q1, q2, b_lengths['q1'], b_lengths['q2'])
    
    if optimizer:
        entropy_loss = nn.CrossEntropyLoss()
        target = Variable(torch.from_numpy(b_data['y'])).cuda()
        optimizer.zero_grad()
        
        # print(target)    
        # print(outputs)
        # print(outputs.data)
        # print(func.softmax(outputs).data)
        
        # loss = func.cross_entropy(outputs, y)
        loss = entropy_loss(outputs, target)
        loss.backward()
        
        optimizer.step()
    
        return loss.item()
        # return loss.data[0]
        # return loss.data
    else:
        _, predicted = outputs.data.max(1)
        prob = func.softmax(outputs).data
        # print(prob)
        # print(predicted)
        return predicted, prob[:, 1]


def run_epoch_eval(dataset, model, output_file=''):
    all_plabels, all_pscores = [], []
    batches, batch_lengths = dataset.get_batches(hparams['batch_size'], ('q1', 'q2', 'y'))
    print(batches)
    print(batch_lengths)
    for b_data, b_lengths in zip(batches, batch_lengths):
        # print(b_data)
        plabels, pscores = run_batch(b_data, b_lengths, model)
        all_plabels.extend(plabels.cpu().numpy().tolist())
        all_pscores.extend(pscores.cpu().numpy().tolist())
    if output_file:
        with open(output_file, 'w') as f:
            for label in all_plabels:
              f.write(f'{label:.6f}\n')
            # for s in all_pscores:
            #     f.write(f'{s:.6f}\n')
    return accuracy_score(dataset.get_data_item('y'), all_plabels)


def run():
    print('Loading data...')
    with open('qdq_emb', 'rb') as f:
      all_sets, embs, word2idx = pickle.load(f)
    
    emb_layer = nn.Embedding(embs.shape[0], embs.shape[1])
    emb_layer.weight = nn.Parameter(torch.from_numpy(embs))
    
    # role_emb_layer = 
    # role_emb_layer.weight =
    # print(emb_layer)
    # print(embs.shape[0])
    # print(embs.shape[1])
    # print(embs)

    model = Model(emb_layer).cuda()
    # .cuda()
    model = model.to(device)

    optimizer = torch.optim.Adam(model.parameters(), lr=hparams['learning_rate'])
    
    train_set = Dataset(all_sets[0], shuffle=True, pad_keys=('q1', 'q2'))
    dev_set = Dataset(all_sets[1], shuffle=False, pad_keys=('q1', 'q2'))
    test_set = Dataset(all_sets[2], shuffle=False, pad_keys=('q1', 'q2'))

    step = 0
    sum_loss = 0
    dev_best = 0
    test_score = 0
    
    print("Starting training...")
    print(hparams)
    
    start_time = time.time()
    
    for epoch in range(hparams['max_epoch']):
    
        batches, batch_lengths = train_set.get_batches(hparams['batch_size'], ('q1', 'q2', 'y'))
  
        for b_data, b_lengths in zip(batches, batch_lengths):  
            sum_loss += run_batch(b_data, b_lengths, model, optimizer)
            step += 1
          
        avg_loss = sum_loss / len(batches)
        sum_loss = 0
        dev_score = run_epoch_eval(dev_set, model)
        out_str = f'Epoch {epoch} iter {step} took {time.time() - start_time:.1f}s\n' \
                  f'loss:\t{avg_loss:.5f}\tdev score:\t{dev_score:.4f}'

        if dev_score > dev_best:
          dev_best = dev_score
          dev_best_epoch = epoch
          # output_file = f'pred/{get_script_short_name(__file__)}.pred'
          test_score = run_epoch_eval(test_set, model, 'DA-BLSTM.txt')
          out_str += f'\t*** New best dev ***\ttest score:\t{test_score:.4f}'
    
        print(out_str)
        start_time = time.time()
        
    print('Best model on dev: dev:{:.4f}\ttest:{:.4f}'.format(dev_best, test_score))
    print('On {:.4f} epoch'.format(dev_best_epoch))

if __name__ == '__main__':
    run()

Loading data...


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:56: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
/usr/local/lib/python3.7/dist-packages/torch/nn/functional.py:1698: UserWarning: nn.functional.tanh is deprecated. Use torch.tanh instead.
  warnings.warn("nn.functional.tanh is deprecated. Use torch.tanh instead.")
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:65: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.


Starting training...
{'learning_rate': 0.001, 'max_epoch': 10, 'display_step': 1000, 'emb_dim': 300, 'conv_win': 3, 'repr_dim': 300, 'fc0_dim': 300, 'fc1_dim': 300, 'n_classes': 2, 'batch_size': 100}
[{'q1': array([[ 1369,   149,  1978, ...,     0,     0,     0],
       [   16,    29,   174, ...,     0,     0,     0],
       [  596,   149,  6634, ...,     0,     0,     0],
       ...,
       [   17, 29748,   350, ...,     0,     0,     0],
       [   16,   177,   924, ...,     0,     0,     0],
       [   51,  2314,    15, ...,     0,     0,     0]]), 'q2': array([[  155,   556,  3065, ...,     0,     0,     0],
       [   16,   660,  3353, ...,     0,     0,     0],
       [    9,  1086,   321, ...,     0,     0,     0],
       ...,
       [ 2714,     1,  2714, ...,     0,     0,     0],
       [  155, 30237, 14973, ...,     0,     0,     0],
       [  155,  2983, 29752, ...,     0,     0,     0]]), 'y': array([0, 0, 0, 0, 0, 1, 0, 1, 1, 0, 1, 1, 1, 0, 0, 1, 0, 1, 0, 1, 0, 0,
       1

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:131: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.


{'q1': array([[  272,    29,   174, ...,     0,     0,     0],
       [  192,     1,   134, ...,     0,     0,     0],
       [15921, 24545,   290, ...,     0,     0,     0],
       ...,
       [   29,  6797,     1, ...,     0,     0,     0],
       [10073,    10,     1, ...,     0,     0,     0],
       [ 1264, 29944, 29945, ...,     0,     0,     0]]), 'q2': array([[    1,   368,   328, ...,    33,  3442,  3443],
       [   33,  2844, 14217, ...,     0,     0,     0],
       [    1,  5659,  5660, ...,     0,     0,     0],
       ...,
       [   29,  6797,     1, ...,     0,     0,     0],
       [  213,  3910,  1138, ...,     0,     0,     0],
       [  873,   302, 30389, ...,     0,     0,     0]]), 'y': array([0, 0, 0, 0, 1, 0, 0, 1, 1, 0, 1, 1, 0, 1, 0, 0, 1, 0, 0, 0, 1, 0,
       1, 1, 0, 0, 0, 0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 0,
       1, 0, 1, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 0, 1, 0, 0, 0, 0, 0, 0,
       0, 1, 0, 0, 1, 0, 1, 0, 0, 1, 1, 0, 0, 0, 1, 0, 1, 0, 0,

## CNN (85.93  baseline accuracy) 

- Best model on dev: dev:0.8358	test:0.8325 

In [ ]:
# !touch CNN.txt
# !touch CNN.pred
# !ls .

import time
import torch
from torch import nn
from torch.nn import functional as func
from torch.autograd import Variable
import pickle
from sklearn.metrics import accuracy_score

hparams = {
    'learning_rate': 0.001,
    'max_epoch': 10,
    'display_step': 1000,
    'emb_dim': 300,
    'conv_win': 3,
    'repr_dim': 300,
    'fc1_dim': 300,
    'n_classes': 2,
    'batch_size': 100
}


class CNN(nn.Module):
    def __init__(self, emb_layer):
        super(CNN, self).__init__()
        self.emb_layer = emb_layer
        self.cnn_layer = nn.Conv1d(hparams['emb_dim'], hparams['repr_dim'], hparams['conv_win'],
                                   padding=int(hparams['conv_win'] / 2))

    def forward(self, x):
        embs = self.emb_layer(x)
        conv = func.tanh(self.cnn_layer(embs.permute(0, 2, 1)))
        out, _ = conv.max(2)
        return out.squeeze()


class Model(nn.Module):
    def __init__(self, emb_layer):
        super(Model, self).__init__()
        self.cnn = CNN(emb_layer)
        self.fc1 = nn.Linear(hparams['repr_dim'] * 2, hparams['fc1_dim'])
        self.fc2 = nn.Linear(hparams['fc1_dim'], hparams['n_classes'])

    def forward(self, q1, q2):
        r1 = self.cnn(q1)
        r2 = self.cnn(q2)
        joint = torch.cat(((r1 - r2).abs(), r1 * r2), dim=1)
        joint = func.tanh(self.fc1(joint))
        out = self.fc2(joint)
        return out


def run_batch(b_data, b_lengths, model, optimizer=None):
    # .cuda()
    q1 = Variable(torch.from_numpy(b_data['q1'])).cuda()
    q2 = Variable(torch.from_numpy(b_data['q2'])).cuda()
    
    outputs = model(q1, q2)
    
    if optimizer:
        entropy_loss = nn.CrossEntropyLoss()
        target = Variable(torch.from_numpy(b_data['y'])).cuda()
        optimizer.zero_grad()
        
        # print(target)    
        # print(outputs)
        # print(outputs.data)
        # print(func.softmax(outputs).data)
        
        # loss = func.cross_entropy(outputs, y)
        loss = entropy_loss(outputs, target)
        loss.backward()
        
        optimizer.step()
    
        return loss.item()
        # return loss.data[0]
        # return loss.data
    else:
        _, predicted = outputs.data.max(1)
        prob = func.softmax(outputs).data
        # print(prob)
        # print(predicted)
        return predicted, prob[:, 1]


def run_epoch_eval(dataset, model, output_file=''):
    all_plabels, all_pscores = [], []
    batches, batch_lengths = dataset.get_batches(hparams['batch_size'], ('q1', 'q2', 'y'))
    for b_data, b_lengths in zip(batches, batch_lengths):
        plabels, pscores = run_batch(b_data, b_lengths, model)
        all_plabels.extend(plabels.cpu().numpy().tolist())
        all_pscores.extend(pscores.cpu().numpy().tolist())
    if output_file:
        with open(output_file, 'w') as f:
            for s in all_pscores:
                f.write(f'{s:.6f}\n')
    return accuracy_score(dataset.get_data_item('y'), all_plabels)


def run():
    print('Loading data...')
    with open('paraphrases_emb', 'rb') as f:
      all_sets, embs, word2idx = pickle.load(f)
    
    emb_layer = nn.Embedding(embs.shape[0], embs.shape[1])
    emb_layer.weight = nn.Parameter(torch.from_numpy(embs))
    
    # role_emb_layer = 
    # role_emb_layer.weight =
    # print(emb_layer)
    # print(embs.shape[0])
    # print(embs.shape[1])
    # print(embs)

    model = Model(emb_layer)
    # .cuda()
    model = model.to(device).cuda()

    optimizer = torch.optim.Adam(model.parameters(), lr=hparams['learning_rate'])
    
    train_set = Dataset(all_sets[0], shuffle=True, pad_keys=('q1', 'q2'))
    dev_set = Dataset(all_sets[1], shuffle=False, pad_keys=('q1', 'q2'))
    test_set = Dataset(all_sets[2], shuffle=False, pad_keys=('q1', 'q2'))

    step = 0
    sum_loss = 0
    dev_best = 0
    test_score = 0
    
    print("Starting training...")
    print(hparams)
    
    start_time = time.time()
    
    for epoch in range(hparams['max_epoch']):
    
        batches, batch_lengths = train_set.get_batches(hparams['batch_size'], ('q1', 'q2', 'y'))
        
        for b_data, b_lengths in zip(batches, batch_lengths):
    
            sum_loss += run_batch(b_data, b_lengths, model, optimizer)
            step += 1
          
        avg_loss = sum_loss / len(batches)
        sum_loss = 0
        dev_score = run_epoch_eval(dev_set, model)
        out_str = f'Epoch {epoch} iter {step} took {time.time() - start_time:.1f}s\n' \
                  f'loss:\t{avg_loss:.5f}\tdev score:\t{dev_score:.4f}'

        if dev_score > dev_best:
          dev_best = dev_score
          dev_best_epoch = epoch
          # output_file = f'pred/{get_script_short_name(__file__)}.pred'
          test_score = run_epoch_eval(test_set, model, 'AI-BLSTM')
          out_str += f'\t*** New best dev ***\ttest score:\t{test_score:.4f}'
    
        print(out_str)
        start_time = time.time()
        
    print('Best model on dev: dev:{:.4f}\ttest:{:.4f}'.format(dev_best, test_score))
    print('On {:.4f} epoch'.format(dev_best_epoch))

if __name__ == '__main__':
    run()

Loading data...


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:56: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
/usr/local/lib/python3.7/dist-packages/torch/nn/functional.py:1698: UserWarning: nn.functional.tanh is deprecated. Use torch.tanh instead.
  warnings.warn("nn.functional.tanh is deprecated. Use torch.tanh instead.")


Starting training...
{'learning_rate': 0.001, 'max_epoch': 10, 'display_step': 1000, 'emb_dim': 300, 'conv_win': 3, 'repr_dim': 300, 'fc1_dim': 300, 'n_classes': 2, 'batch_size': 100}


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:84: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.


Epoch 0 iter 213 took 1.6s
loss:	0.56990	dev score:	0.7134	*** New best dev ***	test score:	0.6995
Epoch 1 iter 426 took 1.6s
loss:	0.40812	dev score:	0.8030	*** New best dev ***	test score:	0.7831
Epoch 2 iter 639 took 1.6s
loss:	0.22850	dev score:	0.8575	*** New best dev ***	test score:	0.8403
Epoch 3 iter 852 took 1.6s
loss:	0.11094	dev score:	0.8455
Epoch 4 iter 1065 took 1.6s
loss:	0.07087	dev score:	0.8631	*** New best dev ***	test score:	0.8551
Epoch 5 iter 1278 took 1.6s
loss:	0.04549	dev score:	0.8775	*** New best dev ***	test score:	0.8723
Epoch 6 iter 1491 took 1.6s
loss:	0.04100	dev score:	0.8887	*** New best dev ***	test score:	0.8707
Epoch 7 iter 1704 took 1.6s
loss:	0.03205	dev score:	0.8751
Epoch 8 iter 1917 took 1.6s
loss:	0.02862	dev score:	0.8831
Epoch 9 iter 2130 took 1.6s
loss:	0.02457	dev score:	0.8711
Best model on dev: dev:0.8887	test:0.8707
On 6.0000 epoch


## AI-CNN (86,55) 
- Best model on dev: dev:0.8401	test:0.8403

In [ ]:
# !touch ai-cnn.txt
# !touch ai-cnn.pred
# !ls .

In [ ]:
import time
import torch
from torch import nn
from torch.nn import functional as func
from torch.autograd import Variable
import pickle
from sklearn.metrics import accuracy_score
# from utils import Dataset, get_script_short_name

hparams = {
    'learning_rate': 0.001,
    'max_epoch': 10,
    'display_step': 1000,
    'emb_dim': 300,
    'conv_win': 3,
    'repr_dim': 300,
    'fc0_dim': 300,
    'fc1_dim': 300,
    'n_classes': 2,
    'batch_size': 100
}


class CNN(nn.Module):
    def __init__(self, emb_layer):
        super(CNN, self).__init__()
        self.emb_layer = emb_layer
        self.cnn_layer = nn.Conv1d(hparams['emb_dim'], hparams['repr_dim'], hparams['conv_win'],
                                   padding=int(hparams['conv_win'] / 2))

    def forward(self, x):
        embs = self.emb_layer(x)
        conv = func.tanh(self.cnn_layer(embs.permute(0, 2, 1)))
        return conv.transpose(1, 2)


class Model(nn.Module):
    def __init__(self, emb_layer):
        super(Model, self).__init__()
        self.cnn = CNN(emb_layer)
        self.fc0 = nn.Linear(hparams['repr_dim'] * 2, hparams['fc0_dim'])
        self.fc1 = nn.Linear(hparams['fc0_dim'] * 2, hparams['fc1_dim'])
        self.fc2 = nn.Linear(hparams['fc1_dim'], hparams['n_classes'])
        self.att = nn.Linear(hparams['repr_dim'] + hparams['fc0_dim'], 1)

    def forward(self, q1, q2, q1_len, q2_len):
       
        r1 = self.cnn(q1)
        r2 = self.cnn(q2)
        h1s, h2s = [], []
       
        for i in range(len(q1_len)):
            r1i = r1[i, :q1_len[i]]
            r2i = r2[i, :q2_len[i]]
        
            r1i_size = r1i.size()
            r2i_size = r2i.size()
            joint_size = torch.Size((r1i_size[0], r2i_size[0], r2i_size[1]))
       
            r1ie = r1i.unsqueeze(1).expand(joint_size)
            r2ie = r2i.unsqueeze(0).expand(joint_size)
       
            joint = torch.cat([(r1ie - r2ie).abs(), r2ie * r1ie], 2)
            joint = func.tanh(self.fc0(joint.view(-1, joint.size()[-1])))
            joint = joint.view(joint_size[0], joint_size[1], hparams['fc0_dim'])
       
            s1, _ = joint.max(1)
            s2, _ = joint.max(0)
       
            s1 = s1.squeeze(1)
            s2 = s2.squeeze(0)

            if (len(s2.shape) == 1):
                s2 = torch.reshape(s2,(1,300))
         
            a1 = self.att(torch.cat([r1i, s1], 1))
            a2 = self.att(torch.cat([r2i, s2], 1))
       
            a1 = func.softmax(a1.t()).t().expand_as(s1)
            a2 = func.softmax(a2.t()).t().expand_as(s2)
       
            h1 = (s1 * a1).sum(0).squeeze()
            h2 = (s2 * a2).sum(0).squeeze()
       
            h1s.append(h1)
            h2s.append(h2)
       
        r1 = torch.stack(h1s)
        r2 = torch.stack(h2s)
       
        joint = torch.cat((r1, r2), dim=1)
        joint = torch.tanh(self.fc1(joint))
       
        out = self.fc2(joint)
        return out


def run_batch(b_data, b_lengths, model, optimizer=None):
    # .cuda()
    q1 = Variable(torch.from_numpy(b_data['q1'])).cuda()
    q2 = Variable(torch.from_numpy(b_data['q2'])).cuda()
    
    outputs = model(q1, q2, b_lengths['q1'], b_lengths['q2'])
    
    if optimizer:
        entropy_loss = nn.CrossEntropyLoss()
        target = Variable(torch.from_numpy(b_data['y'])).cuda()
        optimizer.zero_grad()
        
        # print(target)    
        # print(outputs)
        # print(outputs.data)
        # print(func.softmax(outputs).data)
        
        # loss = func.cross_entropy(outputs, y)
        loss = entropy_loss(outputs, target)
        loss.backward()
        
        optimizer.step()
    
        return loss.item()
        # return loss.data[0]
        # return loss.data
    else:
        _, predicted = outputs.data.max(1)
        prob = func.softmax(outputs).data
        # print(prob)
        # print(predicted)
        return predicted, prob[:, 1]


def run_epoch_eval(dataset, model, output_file=''):
    all_plabels, all_pscores = [], []
    batches, batch_lengths = dataset.get_batches(hparams['batch_size'], ('q1', 'q2', 'y'))
    for b_data, b_lengths in zip(batches, batch_lengths):
        plabels, pscores = run_batch(b_data, b_lengths, model)
        all_plabels.extend(plabels.cpu().numpy().tolist())
        all_pscores.extend(pscores.cpu().numpy().tolist())
    if output_file:
        with open(output_file, 'w') as f:
            for s in all_pscores:
                f.write(f'{s:.6f}\n')
    return accuracy_score(dataset.get_data_item('y'), all_plabels)


def run():
    print('Loading data...')
    with open('paraphrases_emb', 'rb') as f:
      all_sets, embs, word2idx = pickle.load(f)
    
    emb_layer = nn.Embedding(embs.shape[0], embs.shape[1])
    emb_layer.weight = nn.Parameter(torch.from_numpy(embs))
    
    # role_emb_layer = 
    # role_emb_layer.weight =
    # print(emb_layer)
    # print(embs.shape[0])
    # print(embs.shape[1])
    # print(embs)

    model = Model(emb_layer).cuda()
    # .cuda()
    model = model.to(device)

    optimizer = torch.optim.Adam(model.parameters(), lr=hparams['learning_rate'])
    
    train_set = Dataset(all_sets[0], shuffle=True, pad_keys=('q1', 'q2'))
    dev_set = Dataset(all_sets[1], shuffle=False, pad_keys=('q1', 'q2'))
    test_set = Dataset(all_sets[2], shuffle=False, pad_keys=('q1', 'q2'))

    step = 0
    sum_loss = 0
    dev_best = 0
    test_score = 0
    
    print("Starting training...")
    print(hparams)
    
    start_time = time.time()
    
    for epoch in range(hparams['max_epoch']):
    
        batches, batch_lengths = train_set.get_batches(hparams['batch_size'], ('q1', 'q2', 'y'))
        
        for b_data, b_lengths in zip(batches, batch_lengths):
    
            sum_loss += run_batch(b_data, b_lengths, model, optimizer)
            step += 1
          
        avg_loss = sum_loss / len(batches)
        sum_loss = 0
        dev_score = run_epoch_eval(dev_set, model)
        out_str = f'Epoch {epoch} iter {step} took {time.time() - start_time:.1f}s\n' \
                  f'loss:\t{avg_loss:.5f}\tdev score:\t{dev_score:.4f}'

        if dev_score > dev_best:
          dev_best = dev_score
          dev_best_epoch = epoch
          # output_file = f'pred/{get_script_short_name(__file__)}.pred'
          test_score = run_epoch_eval(test_set, model, 'AI-BLSTM')
          out_str += f'\t*** New best dev ***\ttest score:\t{test_score:.4f}'
    
        print(out_str)
        start_time = time.time()
        
    print('Best model on dev: dev:{:.4f}\ttest:{:.4f}'.format(dev_best, test_score))
    print('On {:.4f} epoch'.format(dev_best_epoch))

if __name__ == '__main__':
    run()

Loading data...
Starting training...
{'learning_rate': 0.001, 'max_epoch': 10, 'display_step': 1000, 'emb_dim': 300, 'conv_win': 3, 'repr_dim': 300, 'fc0_dim': 300, 'fc1_dim': 300, 'n_classes': 2, 'batch_size': 100}


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:56: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
/usr/local/lib/python3.7/dist-packages/torch/nn/functional.py:1698: UserWarning: nn.functional.tanh is deprecated. Use torch.tanh instead.
  warnings.warn("nn.functional.tanh is deprecated. Use torch.tanh instead.")
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:79: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:80: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:126: UserWarning: Imp

Epoch 0 iter 213 took 28.6s
loss:	0.56252	dev score:	0.7486	*** New best dev ***	test score:	0.7515
Epoch 1 iter 426 took 28.1s
loss:	0.39190	dev score:	0.8167	*** New best dev ***	test score:	0.8075
Epoch 2 iter 639 took 27.7s
loss:	0.21679	dev score:	0.8535	*** New best dev ***	test score:	0.8455
Epoch 3 iter 852 took 27.7s
loss:	0.12421	dev score:	0.8511
Epoch 4 iter 1065 took 27.7s
loss:	0.08096	dev score:	0.8479
Epoch 5 iter 1278 took 27.5s
loss:	0.05646	dev score:	0.8799	*** New best dev ***	test score:	0.8623
Epoch 6 iter 1491 took 28.3s
loss:	0.03825	dev score:	0.8679
Epoch 7 iter 1704 took 29.2s
loss:	0.03438	dev score:	0.8711
Epoch 8 iter 1917 took 28.8s
loss:	0.03755	dev score:	0.8767
Epoch 9 iter 2130 took 29.1s
loss:	0.03975	dev score:	0.8775
Best model on dev: dev:0.8799	test:0.8623
On 5.0000 epoch


## DA-CNN (84,50 исходн. уровень точности, в статье) - ...

- Best model on dev: dev:0.8251	test:0.8218

- Best model on dev: dev:0.8279	test:0.8265


In [ ]:
# !touch da-cnn.txt
# !touch da-cnn.pred
# !ls .

In [ ]:
import time
import torch
from torch import nn
from torch.nn import functional as func
from torch.autograd import Variable
import pickle
from sklearn.metrics import accuracy_score
# from utils import Dataset, get_script_short_name

hparams = {
    'learning_rate': 0.001,
    'max_epoch': 10,
    'display_step': 1000,
    'emb_dim': 300,
    'conv_win': 3,
    'repr_dim': 300,
    'fc0_dim': 300,
    'fc1_dim': 300,
    'n_classes': 2,
    'batch_size': 100
}


class CNN(nn.Module):
    def __init__(self, emb_layer):
        super(CNN, self).__init__()
        self.emb_layer = emb_layer
        self.cnn_layer = nn.Conv1d(hparams['emb_dim'], hparams['repr_dim'], hparams['conv_win'],
                                   padding=int(hparams['conv_win'] / 2))

    def forward(self, x):
        embs = self.emb_layer(x)
        conv = func.tanh(self.cnn_layer(embs.permute(0, 2, 1)))
        # torch.tanh
        return conv.transpose(1, 2)


class Model(nn.Module):
    def __init__(self, emb_layer):
        super(Model, self).__init__()
        self.cnn = CNN(emb_layer)
        self.fc0 = nn.Linear(hparams['repr_dim'] * 2, hparams['fc0_dim'])
        self.fc1 = nn.Linear(hparams['fc0_dim'] * 2, hparams['fc1_dim'])
        self.fc2 = nn.Linear(hparams['fc1_dim'], hparams['n_classes'])
        self.att = nn.Linear(hparams['repr_dim'] * 2, 1)

    def compare(self, r1, r2, g):
        s = func.softmax(g).unsqueeze(2)
        r2 = r2.unsqueeze(0)
        joint_size = s.size()[:-1] + r2.size()[-1:]
        s = s.expand(joint_size)
        r2 = r2.expand(joint_size)
        ar2 = (s * r2).sum(1).squeeze(1)
        comp = torch.cat([(r1 - ar2).abs(), ar2 * r1], 1)
        comp = func.tanh(self.fc0(comp))
        # torch.tanh
        h = comp.mean(0).squeeze(0)
        return h

    def forward(self, q1, q2, q1_len, q2_len):
        r1 = self.cnn(q1)
        r2 = self.cnn(q2)
        h1s, h2s = [], []
        for i in range(len(q1_len)):
            r1i = r1[i, :q1_len[i]]
            r2i = r2[i, :q2_len[i]]
            r1i_size = r1i.size()
            r2i_size = r2i.size()
            joint_size = torch.Size((r1i_size[0], r2i_size[0], r2i_size[1]))
            r1ie = r1i.unsqueeze(1).expand(joint_size)
            r2ie = r2i.unsqueeze(0).expand(joint_size)
            joint = torch.cat([r1ie, r2ie], 2)
            g = func.tanh(self.att(joint.view(-1, joint.size()[-1])))
            # torch.tanh
            g = g.view(joint_size[0], joint_size[1])
            h1 = self.compare(r1i, r2i, g)
            h2 = self.compare(r2i, r1i, g.t())
            h1s.append(h1)
            h2s.append(h2)
        r1 = torch.stack(h1s)
        r2 = torch.stack(h2s)
        joint = torch.cat((r1, r2), dim=1)
        joint = torch.tanh(self.fc1(joint))
        out = self.fc2(joint)
        return out


def run_batch(b_data, b_lengths, model, optimizer=None):
    # .cuda()
    q1 = Variable(torch.from_numpy(b_data['q1'])).cuda()
    q2 = Variable(torch.from_numpy(b_data['q2'])).cuda()
    
    outputs = model(q1, q2, b_lengths['q1'], b_lengths['q2'])
    
    if optimizer:
        entropy_loss = nn.CrossEntropyLoss()
        target = Variable(torch.from_numpy(b_data['y'])).cuda()
        optimizer.zero_grad()
        
        # print(target)    
        # print(outputs)
        # print(outputs.data)
        # print(func.softmax(outputs).data)
        
        # loss = func.cross_entropy(outputs, y)
        loss = entropy_loss(outputs, target)
        loss.backward()
        
        optimizer.step()
    
        return loss.item()
        # return loss.data[0]
        # return loss.data
    else:
        _, predicted = outputs.data.max(1)
        prob = func.softmax(outputs).data
        # print(prob)
        # print(predicted)
        return predicted, prob[:, 1]


def run_epoch_eval(dataset, model, output_file=''):
    all_plabels, all_pscores = [], []
    batches, batch_lengths = dataset.get_batches(hparams['batch_size'], ('q1', 'q2', 'y'))
    for b_data, b_lengths in zip(batches, batch_lengths):
        plabels, pscores = run_batch(b_data, b_lengths, model)
        all_plabels.extend(plabels.cpu().numpy().tolist())
        all_pscores.extend(pscores.cpu().numpy().tolist())
    if output_file:
        with open(output_file, 'w') as f:
            for s in all_pscores:
                f.write(f'{s:.6f}\n')
    return accuracy_score(dataset.get_data_item('y'), all_plabels)


def run():
    print('Loading data...')
    with open('paraphrases_emb', 'rb') as f:
      all_sets, embs, word2idx = pickle.load(f)
    
    emb_layer = nn.Embedding(embs.shape[0], embs.shape[1])
    emb_layer.weight = nn.Parameter(torch.from_numpy(embs))
    
    # role_emb_layer = 
    # role_emb_layer.weight =
    # print(emb_layer)
    # print(embs.shape[0])
    # print(embs.shape[1])
    # print(embs)

    model = Model(emb_layer).cuda()
    # .cuda()
    model = model.to(device)

    optimizer = torch.optim.Adam(model.parameters(), lr=hparams['learning_rate'])
    
    train_set = Dataset(all_sets[0], shuffle=True, pad_keys=('q1', 'q2'))
    dev_set = Dataset(all_sets[1], shuffle=False, pad_keys=('q1', 'q2'))
    test_set = Dataset(all_sets[2], shuffle=False, pad_keys=('q1', 'q2'))

    step = 0
    sum_loss = 0
    dev_best = 0
    test_score = 0
    
    print("Starting training...")
    print(hparams)
    
    start_time = time.time()
    
    for epoch in range(hparams['max_epoch']):
    
        batches, batch_lengths = train_set.get_batches(hparams['batch_size'], ('q1', 'q2', 'y'))
        
        for b_data, b_lengths in zip(batches, batch_lengths):
    
            sum_loss += run_batch(b_data, b_lengths, model, optimizer)
            step += 1
          
        avg_loss = sum_loss / len(batches)
        sum_loss = 0
        dev_score = run_epoch_eval(dev_set, model)
        out_str = f'Epoch {epoch} iter {step} took {time.time() - start_time:.1f}s\n' \
                  f'loss:\t{avg_loss:.5f}\tdev score:\t{dev_score:.4f}'

        if dev_score > dev_best:
          dev_best = dev_score
          dev_best_epoch = epoch
          # output_file = f'pred/{get_script_short_name(__file__)}.pred'
          test_score = run_epoch_eval(test_set, model, 'AI-BLSTM')
          out_str += f'\t*** New best dev ***\ttest score:\t{test_score:.4f}'
    
        print(out_str)
        start_time = time.time()
        
    print('Best model on dev: dev:{:.4f}\ttest:{:.4f}'.format(dev_best, test_score))
    print('On {:.4f} epoch'.format(dev_best_epoch))

if __name__ == '__main__':
    run()

Loading data...


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:56: VisibleDeprecationWarning: Creating an ndarray from ragged nested sequences (which is a list-or-tuple of lists-or-tuples-or ndarrays with different lengths or shapes) is deprecated. If you meant to do this, you must specify 'dtype=object' when creating the ndarray
/usr/local/lib/python3.7/dist-packages/torch/nn/functional.py:1698: UserWarning: nn.functional.tanh is deprecated. Use torch.tanh instead.
  warnings.warn("nn.functional.tanh is deprecated. Use torch.tanh instead.")
/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:48: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.


Starting training...
{'learning_rate': 0.001, 'max_epoch': 10, 'display_step': 1000, 'emb_dim': 300, 'conv_win': 3, 'repr_dim': 300, 'fc0_dim': 300, 'fc1_dim': 300, 'n_classes': 2, 'batch_size': 100}


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:116: UserWarning: Implicit dimension choice for softmax has been deprecated. Change the call to include dim=X as an argument.


Epoch 0 iter 213 took 34.4s
loss:	0.57643	dev score:	0.7310	*** New best dev ***	test score:	0.7355


KeyboardInterrupt: ignored